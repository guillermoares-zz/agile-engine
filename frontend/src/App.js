import 'bootstrap/dist/css/bootstrap.min.css';
import './App.css'
import React from 'react';
import TransactionAccordion from "./components/transaction-accordion";

function App() {
  return (
    <div className="App">
      <header className="App-header">
        Transaction History
      </header>
      <TransactionAccordion />
    </div>
  );
}

export default App;