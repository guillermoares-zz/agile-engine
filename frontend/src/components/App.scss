.main-card {
    height: 100vh;
}