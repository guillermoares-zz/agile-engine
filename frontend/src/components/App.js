import 'bootstrap/dist/css/bootstrap.min.css';
import './App.scss'
import React, {useEffect, useReducer} from 'react';
import TransactionsView from './transactions-view/transactions-view';
import axios from 'axios';
import {Container} from 'react-bootstrap';
import Card from 'react-bootstrap/Card';
import RefreshButton from './refresh-button/refresh-button';
import StatusInfo, {STATUS_INFO_KIND} from './status-info/status-info';

export const client = axios.create({
  baseURL: `http://localhost:3000`,
  timeout: 5000,
})

export const APP_STATUS = {
  REFRESHING: 'refreshing',
  ERROR: 'error',
  SUCCESS: 'success'
}

export const STATUS_INFO_KIND_BY_APP_STATUS = {
  [APP_STATUS.REFRESHING]: STATUS_INFO_KIND.INFO,
  [APP_STATUS.ERROR]: STATUS_INFO_KIND.ERROR,
  [APP_STATUS.SUCCESS]: STATUS_INFO_KIND.SUCCESS
}

export const APP_ACTION = {
  REFRESH: 'refresh',
  SHOW_ERROR: 'show_error',
  SHOW_TRANSACTIONS: 'show_transactions'
}

export const SUCCESS_MESSAGE = 'Transactions retrieved successfully'
export const REFRESHING_MESSAGE = 'Retrieving transactions...'
export const errorMessage = err => `Error retrieving transactions: ${err.message}`

export const unknownActionMessage = action => `Unknown App action type: ${JSON.stringify(action)}`

export const REFRESHING_STATE = {
  status: APP_STATUS.REFRESHING,
  message: REFRESHING_MESSAGE,
  transactions: []
}

export const INIT_STATE = REFRESHING_STATE

export const isRefreshing = status => status === APP_STATUS.REFRESHING

export const AppReducer = (state, action) => {
  switch (action.type) {
    case APP_ACTION.REFRESH:
      return REFRESHING_STATE
    case APP_ACTION.SHOW_ERROR:
      return {
        status: APP_STATUS.ERROR,
        message: errorMessage(action.error),
        transactions: []
      }
    case APP_ACTION.SHOW_TRANSACTIONS:
      return {
        status: APP_STATUS.SUCCESS,
        message: SUCCESS_MESSAGE,
        transactions: action.transactions
      }
    default:
      throw new Error(unknownActionMessage(action))
  }
}

function App() {
  const [{
    status,
    message,
    transactions
  }, dispatch] = useReducer(AppReducer, INIT_STATE)

  const refreshTransactions = async () => {
    dispatch({type: APP_ACTION.REFRESH})

    try {
      // The delay query parameter forces the backend to delay the response.
      // The idea is to get to see the "loading" state of the component.
      const {data: transactions} = await client.get('/transactions?delay=2000')
      dispatch({
        type: APP_ACTION.SHOW_TRANSACTIONS,
        transactions
      })
    } catch (error) {
      dispatch({
        type: APP_ACTION.SHOW_ERROR,
        error
      })
    }
  }

  useEffect(() => {
    refreshTransactions()
  }, [])

  return (
    <Container fluid="md" className="main-container">
      <Card className="text-center main-card">
        <Card.Header as={'h1'}>
          Transaction History
        </Card.Header>
        <Card.Body>
          <TransactionsView transactions={transactions} isRefreshing={isRefreshing(status)}/>
        </Card.Body>
        <Card.Footer>
          <RefreshButton refreshAction={refreshTransactions} isRefreshing={isRefreshing(status)}/>
          <br/><br/>
          <StatusInfo message={message} kind={STATUS_INFO_KIND_BY_APP_STATUS[status]}/>
        </Card.Footer>
      </Card>
    </Container>
  );
}

export default App;