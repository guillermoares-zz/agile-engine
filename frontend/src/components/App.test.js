import React from 'react';
import {act} from 'react-dom/test-utils'
import App, {
  APP_ACTION,
  APP_STATUS,
  AppReducer,
  errorMessage,
  REFRESHING_STATE,
  SUCCESS_MESSAGE,
  unknownActionMessage,
  client, REFRESHING_MESSAGE
} from './App';
import TransactionsView from './transactions-view/transactions-view';
import {mount, shallow} from 'enzyme';
import StatusInfo, {STATUS_INFO_KIND} from './status-info/status-info';
import RefreshButton from './refresh-button/refresh-button';

describe('App', () => {
  const SAMPLE_TRANSACTIONS = [
    {
      'id': 'e6bea53e-f768-4648-92ec-bdb3ebbdbde1',
      'type': 'credit',
      'amount': 30,
      'effectiveDate': '2020-05-17T15:47:50.643Z'
    },
    {
      'id': '7c75109c-b547-4d1a-b873-e9d7a64c2248',
      'type': 'debit',
      'amount': 23,
      'effectiveDate': '2020-05-17T15:43:23.224Z'
    }
  ]

  test('snapshot renders', () => {
    const wrapper = shallow(<App/>)

    expect(wrapper).toMatchSnapshot()
  })

  describe('Reducer', () => {

    it('should show that the transactions list is refreshing', () => {
      const newState = AppReducer(null, {
        type: APP_ACTION.REFRESH
      })

      expect(newState).toEqual(REFRESHING_STATE)
    })

    it('should show the transactions', () => {
      const newState = AppReducer(null, {
        type: APP_ACTION.SHOW_TRANSACTIONS,
        transactions: SAMPLE_TRANSACTIONS
      })

      expect(newState).toEqual({
        status: APP_STATUS.SUCCESS,
        message: SUCCESS_MESSAGE,
        transactions: SAMPLE_TRANSACTIONS
      })
    })

    it('should show refresh errors', () => {
      const error = new Error('some error message')
      const newState = AppReducer(null, {
        type: APP_ACTION.SHOW_ERROR,
        error
      })

      expect(newState).toEqual({
        status: APP_STATUS.ERROR,
        message: errorMessage(error),
        transactions: []
      })
    })

    it('should detect unknown actions', () => {
      const unknownAction = {type: 'UNKNOWN'}

      expect(() => AppReducer(null, unknownAction))
        .toThrowError(new Error(unknownActionMessage(unknownAction)))
    })
  })

  describe('Rendering', () => {
    beforeEach(() => {
      jest.spyOn(client, 'get')
        .mockImplementation()
    })

    afterEach(() => {
      jest.resetAllMocks()
    })

    it('renders the TransactionsView', () => {
      let wrapper = shallow(<App/>);
      expect(wrapper.find(TransactionsView).length).toEqual(1);
    });

    it('renders the StatusInfo', () => {
      let wrapper = shallow(<App/>);
      expect(wrapper.find(StatusInfo).length).toEqual(1);
    });

    it('renders the RefreshButton', () => {
      let wrapper = shallow(<App/>);
      expect(wrapper.find(RefreshButton).length).toEqual(1);
    });

    it('retrieves transactions on init', async () => {
      await act(async () => {
        mount(<App/>);
      });
      expect(client.get).toBeCalledWith('/transactions?delay=2000')
    })

    it('shows as refreshing at initialization', async () => {
      let wrapper

      await act(async () => {
        wrapper = mount(<App/>);
      })

      expect(wrapper.find(StatusInfo).props()).toEqual({
        message: REFRESHING_MESSAGE,
        kind: STATUS_INFO_KIND.INFO
      })

      expect(wrapper.find(TransactionsView).props()).toEqual({
        isRefreshing: true,
        transactions: []
      })

      expect(wrapper.find(RefreshButton).props().isRefreshing).toBe(true)
      expect(wrapper.find(RefreshButton).props().refreshAction).toBeDefined()
    })

    it('shows as refreshing while transactions are being retrieved', async () => {
      client.get = () => new Promise(res => setInterval(res, 2000))

      let wrapper

      await act(async () => {
        wrapper = mount(<App/>);
      })

      wrapper.update()

      expect(wrapper.find(StatusInfo).props()).toEqual({
        message: REFRESHING_MESSAGE,
        kind: STATUS_INFO_KIND.INFO
      })

      expect(wrapper.find(TransactionsView).props()).toEqual({
        isRefreshing: true,
        transactions: []
      })

      expect(wrapper.find(RefreshButton).props().isRefreshing).toBe(true)
      expect(wrapper.find(RefreshButton).props().refreshAction).toBeDefined()
    })

    it('shows retrieved transactions', async () => {
      client.get = () => ({data: SAMPLE_TRANSACTIONS})

      let wrapper

      await act(async () => {
        wrapper = mount(<App/>);
      })

      wrapper.update()

      expect(wrapper.find(StatusInfo).props()).toEqual({
        message: SUCCESS_MESSAGE,
        kind: STATUS_INFO_KIND.SUCCESS
      })

      expect(wrapper.find(TransactionsView).props()).toEqual({
        isRefreshing: false,
        transactions: SAMPLE_TRANSACTIONS
      })

      expect(wrapper.find(RefreshButton).props().isRefreshing).toBe(false)
      expect(wrapper.find(RefreshButton).props().refreshAction).toBeDefined()
    })

    it('shows errors while retrieving transactions', async () => {
      const error = new Error('Some error occurred')
      client.get = () => {
        throw error
      }

      let wrapper

      await act(async () => {
        wrapper = mount(<App/>);
      })

      wrapper.update()

      expect(wrapper.find(StatusInfo).props()).toEqual({
        message: errorMessage(error),
        kind: STATUS_INFO_KIND.ERROR
      })

      expect(wrapper.find(TransactionsView).props()).toEqual({
        isRefreshing: false,
        transactions: []
      })

      expect(wrapper.find(RefreshButton).props().isRefreshing).toBe(false)
      expect(wrapper.find(RefreshButton).props().refreshAction).toBeDefined()
    })
  })
})
