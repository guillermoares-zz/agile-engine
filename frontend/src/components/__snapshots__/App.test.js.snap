// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<Container
  className="main-container"
  fluid="md"
>
  <Card
    body={false}
    className="text-center main-card"
  >
    <CardHeader
      as="h1"
    >
      Transaction History
    </CardHeader>
    <CardBody>
      <TransactionsView
        isRefreshing={true}
        transactions={Array []}
      />
    </CardBody>
    <CardFooter>
      <RefreshButton
        isRefreshing={true}
        refreshAction={[Function]}
      />
      <br />
      <br />
      <StatusInfo
        kind="info"
        message="Retrieving transactions..."
      />
    </CardFooter>
  </Card>
</Container>
`;
